src/Compilers/Core/Portable/SymbolDisplay/AbstractSymbolDisplayVisitor.cs

```diff
@@ -75,6 +68,53 @@ protected AbstractSymbolDisplayVisitor NotFirstVisitorNamespaceOrType
             }
         }
 
+        protected virtual void Allocate(
+            ArrayBuilder<SymbolDisplayPart> builder,
+            SymbolDisplayFormat format,
+            bool isFirstSymbolVisited,
+            SemanticModel? semanticModelOpt,
+            int positionOpt,
+            bool inNamespaceOrType)
+        {
+            Debug.Assert(format != null);
+
+            if (_format is not null)
+                throw new InvalidOperationException();
+
+            this._builder = builder;
+            this._format = format;
+            this._isFirstSymbolVisited = isFirstSymbolVisited;
+
+            this._semanticModelOpt = semanticModelOpt;
+            this._positionOpt = positionOpt;
+            this._inNamespaceOrType = inNamespaceOrType;
+
+            // If we're not the first symbol visitor, then we will just recurse into ourselves.
+            if (!isFirstSymbolVisited)
+            {
+                _lazyNotFirstVisitor = this;
+            }
+        }
+
+        public virtual void Free()
+        {
+            if (_lazyNotFirstVisitor != this)
+                _lazyNotFirstVisitor?.Free();
```

Comment:
>_lazyNotFirstVisitor [](http://example.com/codeflow?start=16&length=20)

Since we do not control how this visitor is created (it looks like `MakeNotFirstVisitor` does), it is not appropriate to free the visitor. Instead, whoever implements `MakeNotFirstVisitor`, should override `Free` accordingly.

Reply:
> whoever implements ```MakeNotFirstVisitor```, should override Free accordingly. 

On the second thought. I think a better way to deal with this is to add 
```
protected abstract void FreeNotFirstVisitor(AbstractSymbolDisplayVisitor notFirstVisitor, bool inNamespaceOrType = false);
```
and call it from here.

Reply:
➡️ Now updated